SWATCH-2306: Create grafana dashboard for PAYG metrics

Sgitario opened the pull request:
Jira issue: SWATCH-2306

## Description
The metrics to use for each are:
- Metered, which is the first "phase". It's swatch_metrics_ingested_usage_total: Added by SWATCH-2297
- Tallied, which is the second one: It's swatch_tally_tallied_usage_total: Added by SWATCH-2299
- Covered by contract: It's swatch_contract_usage_total : Added by SWATCH-2300
- Billing pending: It's swatch_billable_usage_total Added by SWATCH-2301
- Remitted: It`s swatch_producer_metered_total  Added by SWATCH-2302

The new "Metered end-to-end stats" panels look like as:

![Screenshot From 2025-02-07 10-29-21](https://github.com/user-attachments/assets/d484eaa6-8869-44b5-a8da-7391352b23b4)

Which displays the values by the select product(s), metric and billing provider.

As part of this task, I found the following issues:
- The metric "swatch_billable_usage_total" is being counting twice: SWATCH-3294
- The metric "swatch_tally_tallied_usage_total" is always empty. I will dig further into this issue.

Additionally to the above panels, I've also added more panels to see this data by metric:

![Screenshot From 2025-02-07 10-33-23](https://github.com/user-attachments/assets/e3956224-85c6-4fde-955d-32e58316a789)

The main difference is that we can see the number for all the metrics at once without needing to pick up one concrete metric.

## Testing
Extract the new dashboard json using `oc extract -f .rhcicd/grafana/grafana-dashboard-subscription-watch-payg-metrics.configmap.yaml --confirm`

Once you extract it from the .yaml that's checked into this repo, you can import it into the stage instance of grafana by going to `Dashboards -> +Import` from the left nav.

barnabycourt commented:
Is there a way to create the panels such that we can hover with the mouse on each day and see the value? Also, is it worth breaking out to have a panel for each product/metric instead of having to select them from the drop-down?

Sgitario commented:
> Is there a way to create the panels such that we can hover with the mouse on each day and see the value? 

I like this idea. Wdyt @kahowell ?

> Also, is it worth breaking out to have a panel for each product/metric instead of having to select them from the drop-down?

This is why I added an additional panel which is by product and metric (see the second screenshot from the PR description). In this case, selecting the metric from the drop-down will have no effects.

kahowell reviewed (changes requested):
- [x] Instead of specifying a hard-coded range selector (`[1d]`), it's more useful to use the grafana global variable `$__range`, because then the time selection can be used to choose a specific time range (both the ending timestamp and the time period of the range).
- [ ] Instead of hard-coding the `product`, `metric_id`, and `billing_provider` values, define them as `Query` variables, using `${datasource}` and the metrics ingested metric e.g. (`label_values(swatch_metrics_ingested_usage_total,product)`).

kahowell commented:
> > Is there a way to create the panels such that we can hover with the mouse on each day and see the value?

> I like this idea. Wdyt @kahowell ?

[You'll have to switch to the time series panel type for that](https://stackoverflow.com/questions/70832216/grafana-show-value-at-time-index-in-stat-graph-on-hover). I'm fine with that.

kartikshahc commented:
@Sgitario  I might be missing something here because none of the dashboard metrics meet with what we are getting from database. So please correct me if I am understanding these metrics correctly as these were mostly implemented while I was out in Dec.

Question: The metrics shown in dashboard is for 1 day or for the month?

Example for Rosa on stage for 1 day period:
For Cores metric I see 8 failed status but, dashboard says 533

Gabi result for 1 day and for this month below screenshot:
![Screenshot From 2025-02-06 16-29-19](https://github.com/user-attachments/assets/1dfc6977-0556-4609-9890-f1c5cdf5fddf)


Dashboard screenshot:
![Screenshot From 2025-02-06 16-28-01](https://github.com/user-attachments/assets/67776c09-cd7a-4b71-b992-4e519fb7d336)

Sgitario commented:
> @Sgitario I might be missing something here because none of the dashboard metrics meet with what we are getting from database. So please correct me if I am understanding these metrics correctly as these were mostly implemented while I was out in Dec. Also this issue could be with the cards implemented for capturing metrics and not with this card but, just wanted to bring it up.
> 
> **Question: The metrics shown in dashboard is for 1 day or for the month?**
> 
> Example for Rosa on stage for 1 day period: For Cores metric I see 8 failed status but, dashboard says 533
> 
> Gabi result for 1 day and for this month below screenshot: `gabi "select sum(remitted_pending_value),status,metric_id from billable_usage_remittance where product_id='rosa' and remittance_pending_date>'2025-02-05' group by status, metric_id"`
> 
> ![Screenshot From 2025-02-06 16-29-19](https://private-user-images.githubusercontent.com/11471167/410659121-1dfc6977-0556-4609-9890-f1c5cdf5fddf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg5MDIwOTcsIm5iZiI6MTczODkwMTc5NywicGF0aCI6Ii8xMTQ3MTE2Ny80MTA2NTkxMjEtMWRmYzY5NzctMDU1Ni00NjA5LTk4OTAtZjFjNWNkZjVmZGRmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA3VDA0MTYzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTliMTYwYWU1YmUxZWFiN2YwNzc5OGEzMzJlYTUxYzJjZjg0YTJmOTkxZTA5NTYyNzVlZTYzNWFlY2ZiOTYyZGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._ylPbCb8ECA8sCJjAAwXbbAzG6TQLfUKSjVJ6oxCjMY)
> 
> Dashboard screenshot: ![Screenshot From 2025-02-06 16-28-01](https://private-user-images.githubusercontent.com/11471167/410658368-67776c09-cd7a-4b71-b992-4e519fb7d336.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg5MDIwOTcsIm5iZiI6MTczODkwMTc5NywicGF0aCI6Ii8xMTQ3MTE2Ny80MTA2NTgzNjgtNjc3NzZjMDktY2Q3YS00YjcxLWI5OTItNGU1MTlmYjdkMzM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA3VDA0MTYzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFjNDg2YWJjOTJhNTUzMzdlNjNkYWRiMjNlZTgxNTAzMTUwNTQ2N2RjYjEyZDc2Yjg2NmQzYTg1ODgyMGU1MTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Go4BsTEhv97VsaKN9vw4svdeghlFWu50cOyQTGSS_CM)
> 
> If you notice none of the values in billing pending or remitted matches with the database values ![Screenshot From 2025-02-06 16-34-06](https://private-user-images.githubusercontent.com/11471167/410661893-c5011736-72ed-4e8c-814c-f43bb585618b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg5MDIwOTcsIm5iZiI6MTczODkwMTc5NywicGF0aCI6Ii8xMTQ3MTE2Ny80MTA2NjE4OTMtYzUwMTE3MzYtNzJlZC00ZThjLTgxNGMtZjQzYmI1ODU2MThiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA3VDA0MTYzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk1YTRhZDllYjg1NzM5NTFkMTBlY2YzZDg4YWUwNWMwMzQ4MTExZjEyNTVhMGJjMjQ5NmFmMjBhYjZlODQ5ZTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.H6tgiQNyRdQ9X4eGT94BDV7XCx_MogTt0R5mBwxfKoo)
> 
> **Same applies for metered panel:** Gabi screenshot below and query shows 2464 Instance hours for the day Feb 6 and 4543 for this month but, doesn't match the dashboard value 5702. `gabi "select sum((data->'measurements'->0->'value')::float),data->'product_tag', data->'measurements'->0->>'metric_id' from events where timestamp>='2025-02-06T00:00:00Z' group by data->'product_tag', data->'measurements'->0->>'metric_id'"`
> 
> ![Screenshot From 2025-02-06 16-51-24](https://private-user-images.githubusercontent.com/11471167/410668351-89a10d47-a510-4e29-9849-09c086e32b45.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg5MDIwOTcsIm5iZiI6MTczODkwMTc5NywicGF0aCI6Ii8xMTQ3MTE2Ny80MTA2NjgzNTEtODlhMTBkNDctYTUxMC00ZTI5LTk4NDktMDljMDg2ZTMyYjQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA3VDA0MTYzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVjNmUwODY4OWE1MTFjOTA5YTAxN2U1OTg3YmZjMWExNjY0N2Q1NGYzMWNlNmE0YmI5MDgyOThiZTA3ZjExODMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bqnT-T-CVlI09SH8QnvPvghJeP4U76ZLH0olqI35ldk)
> 
> Dashboard results: ![Screenshot From 2025-02-06 16-54-50](https://private-user-images.githubusercontent.com/11471167/410668648-fa8a8e11-0982-42c8-ad21-c5be33b0f6f9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg5MDIwOTcsIm5iZiI6MTczODkwMTc5NywicGF0aCI6Ii8xMTQ3MTE2Ny80MTA2Njg2NDgtZmE4YThlMTEtMDk4Mi00MmM4LWFkMjEtYzViZTMzYjBmNmY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA3VDA0MTYzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdjNGZlY2E0ZTIyNTI4ZjJjMTQyYjZiMjU4NzFkOTQ3NmQwZDkwOGY2YWQ2MjYyNDg3MDc4MGUzYzM3MmU0YWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.o4_uu6a1ydJJu8RUJY9oZyRxYdua7tGZUYE52d9CROA)

Yes, the numbers are very wrong that's why I started digging into this and found issues like double or triple counting some metrics, incremeting the counters using different units or using a different format in the counters. 

Still, thank you very much for the thorough review! This is really valuable. After addressing all these issues in https://github.com/RedHatInsights/rhsm-subscriptions/pull/4159 and having it merged in stage, I will update the dashboard and update it after the feedback from Kevin and Barnaby, and verify that everything look good.

Sgitario commented:
@barnabycourt @kahowell @kartikshahc , I've updated the dashboard with the following changes:
- use the grafana global variable $__range for all the panels instead of the hardcoded range of "1d"
- Instead of hard-coding the product, metric_id, and billing_provider values, define them as Query variables, using ${datasource} and the metrics ingested metric
- changed the panels to use timed series, so we'll see the plots by days

I've updated the screenshots in the PR description.

Note @kartikshahc that the numbers are not trustable yet because https://github.com/RedHatInsights/rhsm-subscriptions/pull/4159 is not merged yet.

InsightsDroid commented:
### IQE Tests Summary Report
- tests image: https://quay.io/cloudservices/iqe-tests:rhsm-subscriptions

kartikshahc commented:
> @barnabycourt @kahowell @kartikshahc , I've updated the dashboard with the following changes:
> 
> * use the grafana global variable $__range for all the panels instead of the hardcoded range of "1d"
> * Instead of hard-coding the product, metric_id, and billing_provider values, define them as Query variables, using ${datasource} and the metrics ingested metric
> * changed the panels to use timed series, so we'll see the plots by days
> * additionally, I've included a couple of new panels to check the remittance plots by failure unknown and failure subscription not found
> 
> I've updated the screenshots in the PR description.
> 
> Note @kartikshahc that the numbers are not trustable yet because #4159 is not merged yet.

yea we should probably wait for #4159 to be merged first since the datapoints seems weird. I still don't see the graphs plotted correctly like the "Billing pending" for rosa in top graph and bottom graph is different. The top graph doesn't show any value in "Billing pending" and bottom graph does have it. 
Additionally, I noticed some of the products are not yet shown in graph like `rhel-for-x86-els-payg`.
I checked the changes you mentioned above and those seem fine to me.

Sgitario commented:
> > > @barnabycourt @kahowell @kartikshahc , I've updated the dashboard with the following changes:
> > > 
> > > * use the grafana global variable $__range for all the panels instead of the hardcoded range of "1d"
> > > * Instead of hard-coding the product, metric_id, and billing_provider values, define them as Query variables, using ${datasource} and the metrics ingested metric
> > > * changed the panels to use timed series, so we'll see the plots by days
> > > * additionally, I've included a couple of new panels to check the remittance plots by failure unknown and failure subscription not found
> > > 
> > > I've updated the screenshots in the PR description.
> > > Note @kartikshahc that the numbers are not trustable yet because #4159 is not merged yet.
> > 
> > 
> > yea we should probably wait for #4159 to be merged first since the datapoints seems weird. I still don't see the graphs plotted correctly like the "Billing pending" for rosa in top graph and bottom graph is different. The top graph doesn't show any value in "Billing pending" and bottom graph does have it. Additionally, I noticed some of the products are not yet shown in graph like `rhel-for-x86-els-payg`. I checked the changes you mentioned above and those seem fine to me.
> 
> @Sgitario overall graph legends and what you are planning to show seems good. However the data still seems incorrect after you merged #4159. Also as I mentioned some products are not showing up in graph. Let me know if you are looking for any other information from me? I can approve these changes if you are planning to create a new card to address them.

I merged #4159 this morning, so I expect the data starts looking good tomorrow (after the tests in stage are executed). About the products not showing up in the graph, this strongly depends on the datasources, so if one product is missing, it's because there is no data in stage/prod.

InsightsDroid commented:
### IQE Tests Summary Report
- tests image: https://quay.io/cloudservices/iqe-tests:rhsm-subscriptions

Sgitario commented:
@kahowell @kartikshahc the data looks much better now if you filter the data using the last 12 hours. Can you take another look for the PR review?
